hilti/toolchain/src/compiler/cfg.cc

```diff
@@ -0,0 +1,893 @@
+                                                            return escape(x->print());
+                                                        });
+
+                                                        std::sort(xs.begin(), xs.end());
+                                                        return xs;
+                                                    }(),
+                                                    [](auto&& x) { return ! x.empty(); }),
+                                                ", "));
+                });
+                std::sort(xs.begin(), xs.end());
+                if ( ! xs.empty() )
+                    return util::fmt("kill: [%s]", util::join(xs, " "));
+                else
+                    return std::string();
+            }();
+
+            auto reachability = [&]() -> std::string {
+                auto&& r = transfer.reachability;
+                if ( ! r )
+                    return "";
+
+                auto to_str = [&](auto&& xs) {
+                    auto ys = util::transformToVector(xs, [&](auto&& x) { return escape(x->print()); });
+                    std::sort(ys.begin(), ys.end());
+                    return util::join(ys, ", ");
+                };
+
+                return util::fmt("reach: { in: [%s] out: [%s] }", to_str(r->in), to_str(r->out));
+            }();
+
+            auto aliases = [&]() {
+                auto xs = util::transformToVector(transfer.aliases, [&](auto* decl) {
+                    return escape(decl->template as<const hilti::Declaration>()->id());
+                });
+                std::sort(xs.begin(), xs.end());
+                if ( ! xs.empty() )
+                    return util::fmt("aliases: [%s]", util::join(xs, ", "));
+                else
+                    return std::string();
+            }();
+
+            auto keep = [&]() -> std::string { return transfer.keep ? "keep" : ""; }();
+
+            xlabel = util::fmt("xlabel=\"%s\"",
+                               util::join(util::filter(std::vector{use, gen, kill, reachability, aliases, keep},
+                                                       [](auto&& x) { return ! x.empty(); }),
+                                          " "));
+        }
+
+        if ( auto&& meta = n->tryAs<MetaNode>() ) {
+            if ( meta->isA<Start>() )
+                ss << util::fmt("    %s [label=start shape=Mdiamond %s];\n", id, xlabel ? *xlabel : "");
+
+            else if ( meta->isA<End>() )
+                ss << util::fmt("    %s [label=end shape=Msquare %s];\n", id, xlabel ? *xlabel : "");
+
+            else if ( meta->isA<Flow>() )
+                ss << util::fmt("    %s [shape=point %s];\n", id, xlabel ? *xlabel : "");
+
+            else if ( auto* scope = meta->tryAs<ScopeEnd>() ) {
+                ss << util::fmt("    %s [label=\"scope_end %s\" shape=triangle %s];\n", id,
+                                scope->scope->meta().location(), xlabel ? *xlabel : "");
+            }
```

Comment:
Not having a `ScopeBegin` is a readability nightmare for these CFGs with shadowed variables :(

```
module Test {

import hilti;

function void f() {
    local x = 5;
    hilti::print(x);
    {
        local x = 6;
        hilti::print(x);
        {
            local x = 7;
            hilti::print(x);

        }
    }
}

f();

}
```
![cfg](https://github.com/user-attachments/assets/f975b633-28c5-4eec-a208-178e34726317)